Update test_runlib.py

IshaVDave opened the pull request:
Fixes issue in-toto#365:

Now checks the platform of the user's system. If it's windows, it raises an error that asks the user if the developer mode is activated.

lukpueh commented on tests/test_runlib.py:
```diff
@@ -1,22 +1,18 @@
+
```
Please don't add or remove blank lines that are not related to the lines you are working on.

lukpueh commented on tests/test_runlib.py:
```diff
@@ -278,99 +274,111 @@ def test_record_dot_check_files_hash_dict_schema(self):
 
+  def raiseException():
+    if os.name == "nt":
+      raise IOError("Is developer mode enabled?")
```
Some notes on [code style](https://github.com/secure-systems-lab/code-style-guidelines) and programming style: 
- We usually use snake_case and not camelCase for function names.
- We often start small helper functions like this with an underscore.
- Please try to choose mnemonic function names. I would maybe go for something like `_raise_win_dev_mode_error`.
- Please always add a docstring, to your functions, for short functions like this one it can be a oneliner.
- I think the error message could need some additional context. If I saw this in my error log, I would like to know why this IOError asks me about developer mode. :)
- It might be helpful (also for context) to include the error message from the original exception.

lukpueh reviewed (changes requested):
Congrats on your first PR, @IshaDave! 🎉 This looks quite good to me. I have left a couple of suggestions inline. Please address them and I'll give it another look.

On a more general note, it would be great if you could write [more meaningful commit messages](https://github.com/secure-systems-lab/lab-guidelines/blob/master/commits.md#how-to-write-good-commit-messages).

IshaVDave commented:
Thank you Lukas! I was also surprised about the deleted lines because I
hadn't done it intentionally. I'll fix that and implement all the code
style and context-related changes ASAP :)

On Mon, Jun 8, 2020 at 9:03 PM lukpueh <notifications@github.com> wrote:

> *@lukpueh* requested changes on this pull request.
>
> Congrats on your first PR, @IshaDave
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_IshaDave&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=DpS_tBZA8f8Ds55RqOhkTS4KjMav4cfiUIV6EChD39U&e=>!
> 🎉 This looks quite good to me. I have left a couple of suggestions
> inline. Please address them and I'll give it another look.
>
> On a more general note, it would be great if you could write more
> meaningful commit messages
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_secure-2Dsystems-2Dlab_lab-2Dguidelines_blob_master_commits.md-23how-2Dto-2Dwrite-2Dgood-2Dcommit-2Dmessages&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=RMArvk4QuZe7Vb2XBd3uJ8qNhTPBUVJvPRT9wB2XFNo&e=>
> .
> ------------------------------
>
> In tests/test_runlib.py
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_in-2Dtoto_in-2Dtoto_pull_368-23discussion-5Fr436780044&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=VLFqBoGhwJlJ12wKGk2vFfSd8Wtk4je_HVJNE2ap-m4&e=>
> :
>
> > @@ -1,22 +1,18 @@
> +
>
> Please don't add or remove blank lines that are not related to the lines
> you are working on.
> ------------------------------
>
> In tests/test_runlib.py
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_in-2Dtoto_in-2Dtoto_pull_368-23discussion-5Fr436796615&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=mgmN6msHKItyGgjnZgB2P667xpRDHck6YlmWVjCthQU&e=>
> :
>
> > @@ -278,99 +274,111 @@ def test_record_dot_check_files_hash_dict_schema(self):
>        sorted(self.full_file_path_list))
>
>
> +  def raiseException():
> +    if os.name == "nt":
> +      raise IOError("Is developer mode enabled?")
>
> Some notes on code style
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_secure-2Dsystems-2Dlab_code-2Dstyle-2Dguidelines&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=VBAwKHs4SZW6T7otk8LhOFY5AQi749t7pfrY9_dV3wg&e=>
> and programming style:
>
>    - We usually use snake_case and not camelCase for function names.
>    - We often start small helper functions like this with an underscore.
>    - Please try to choose mnemonic function names. I would maybe go for
>    something like _raise_win_dev_mode_error.
>    - Please always add a docstring, to your functions, for short
>    functions like this one it can be a oneliner.
>    - I think the error message could need some additional context. If I
>    saw this in my error log, I would like to know why this IOError asks me
>    about developer mode. :)
>    - It might be helpful (also for context) to include the error message
>    from the original exception.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_in-2Dtoto_in-2Dtoto_pull_368-23pullrequestreview-2D426336671&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=tctJSsZJ0C2eO0R0Q6D3BH4M6vUOiDCiCfIWdPcJwNc&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AO7YDQDP622NTVMJBEEF4ZTRVUADJANCNFSM4NYKCPPA&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=N7r6kqjZPrcT66vpuZq6ww02y61BCQkyysajFYSEnPk&s=Xj0k3UubPiWFZmktwQ-_mT4LqadK4E9t8QMHLFuFFcs&e=>
> .
>

lukpueh commented:
Thanks, @IshaDave!

> I was also surprised about the deleted lines because I hadn't done it intentionally

I suspect it was your sneaky IDE. ;)  As a general rule, please always make sure that the [diff](https://github.com/in-toto/in-toto/pull/368/files) of your PRs only contain the changes you intended.

IshaVDave commented:
Hi @lukpueh. I implemented all the edits, let me know if you feel something else needs to be changed too!

lukpueh commented:
Thank you, @IshaDave! Did you take a look at the changes your PR contains, as I asked above? It seems to me that there are still unintended edits. :P

Also, we perform automatic checks on PRs and require them to pass before we merge. As you can see in the table below "Some checks were not successful". Can you please try to find out what went wrong and fix it?

IshaVDave commented:
Okay, sure. On it!

On Tue, Jun 9, 2020 at 2:59 PM lukpueh <notifications@github.com> wrote:

> Thank you, @IshaDave
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_IshaDave&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=OWwfGjT8xiaYaBqlIZq1oF3EiORCnWlRTvlI3m8Vr-Y&s=-7lXUxk-bYZlwgXiyOmONcBZ028tbeYzqHqB2EJswko&e=>!
> Did you take a look at the changes your PR contains, as I asked above? It
> seems to me that there are still unintended edits. :P
>
> Also, we perform automatic checks on PRs and require them to pass before
> we merge. As you can see in the table below "Some checks were not
> successful". Can you please try to find out what went wrong and fix it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_in-2Dtoto_in-2Dtoto_pull_368-23issuecomment-2D641159174&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=OWwfGjT8xiaYaBqlIZq1oF3EiORCnWlRTvlI3m8Vr-Y&s=C3l-UYbDS5P1bX3GNYUj3tYcC_dmILI6AcRwUGWhAww&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AO7YDQBTY25S5I3ZRKQU27LRVX6G3ANCNFSM4NYKCPPA&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=M8MLsgsebbpS89V64OzkoQ&m=OWwfGjT8xiaYaBqlIZq1oF3EiORCnWlRTvlI3m8Vr-Y&s=OLjDf5OZY8WZbBfri951Rd4bgkpST9R8S4Q9iGBGpGQ&e=>
> .
>

IshaVDave committed:
Add windows dev mode hint to symlink tests

Added an extra method '_raise_win_dev_mode_error' to the 'Test_ApplyExcl
udePatterns' class in test_runlib.py, and updated 3 other methods to inc
lude it . It checks if the user is on a Windows platform. If so, they ar
e informed that Developer mode is required to work with symlinks on Wind
ows, and asked if it is enabled.

Signed-off-by: Isha Dave <ishavdave@nyu.edu>

lukpueh commented on tests/test_runlib.py:
```diff
@@ -278,99 +278,111 @@ def test_record_dot_check_files_hash_dict_schema(self):
 
 
+  def _raise_win_dev_mode_error(self):
+    # If the platform is Windows, raises an error that asks the user if developer mode is activated.
```
Minor nit: This is a comment (`# ....`) but should be a docstring (`"""...."""`).

lukpueh commented on tests/test_runlib.py:
```diff
@@ -278,99 +278,111 @@ def test_record_dot_check_files_hash_dict_schema(self):
+      for pair in link_pairs:
+        os.unlink(pair[1])
+    except IOError:
+      TestRecordArtifactsAsDict._raise_win_dev_mode_error(self)
```
Minor nit: I suggest this syntax `self. _raise_win_dev_mode_error()` when calling instance methods. It just feels more idiomatic. I tend to only use the `ClassName.method_name`-syntax when calling class- or static-methods.

By the way, since you don't need the method nor the class instance in `_raise_win_dev_mode_error` you should consider making it a `staticmethod`. :)

@SantiagoTorres, @adityasaky, what do you think?

IshaVDave replied:
I agree. I initially had planned to make it static, but decided instead to make it uniform with all the other methods. Now that you voice my opinion too, I think I'll just go with static :)

lukpueh replied:
Regarding uniformity, the difference is that `_raise_win_dev_mode_error` is a pure utility function with no ties to the class or its instances, whereas the other methods do need to be instance methods to be executed as `unittest`.

lukpueh reviewed (changes requested):
Thanks for fixing the git history, @IshaDave! I have left to minor comments inline and I also have one general concern about the design here. Correct me if I'm wrong, but I am pretty sure that if any of the code in your try block(s) raises an `IOError` and the platform is not Windows, then the test would still pass, which is not what we want.

The issue is amplified by the fact that there is a lot of code in the try block, so there are a lot of places where an unexpected error could be raised. It generally is good programming style to only wrap the part in try/else that is expected to raise an exception.

I think you should do the following:
1. Wrap only the line that you expect to raise an IOError on Windows in a `try` block. (It should be enough to only wrap the first such line, because if there is a problem, the test should not get past it anyways).
2. In the corresponding `except` block you check
    - If you're indeed on Windows, and raise the exception with the developer mode hint (as you already do),
    - Otherwise, if you are not on Windows, the exception comes fully unexpected and we sure want to know about it. This means the exception must be re-`raise`ed. :)

What do you think?

IshaVDave commented:
> I think you should do the following:
> 
> 1. Wrap only the line that you expect to raise an IOError on Windows in a `try` block. (It should be enough to only wrap the first such line, because if there is a problem, the test should not get past it anyways).
> 2. In the corresponding `except` block you check
>    
>    * If you're indeed on Windows, and raise the exception with the developer mode hint (as you already do),
>    * Otherwise, if you are not on Windows, the exception comes fully unexpected and we sure want to know about it. This means the exception must be re-`raise`ed. :)
> 
> What do you think?

Yes, I realize what you pointed out. I had this conversation with @SantiagoTorres too, but I missed on incorporating the 're-raising'. I'll do that, thanks a lot! :)

lukpueh commented on tests/test_runlib.py:
```diff
@@ -277,6 +277,12 @@ def test_record_dot_check_files_hash_dict_schema(self):
+  def _raise_win_dev_mode_error():
+    """If the platform is Windows, raises an error that asks the user if developer mode is activated."""
+    if os.name == "nt":
+      raise IOError("Developer mode is required to work with symlinks on Windows. Is it enabled?")
```
Looks like we [don't flag line length in the tests](https://github.com/in-toto/in-toto/blob/41878e286eafceba41dac8d1e352eff1c7c44713/tox.ini#L29-L31). Could you still please wrap the docstring and the error message at 80 chars? Thanks. :)

lukpueh reviewed (approved):
Excellent work, @IshaDave!
